File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let longest = 0;
    let left = 0;
    const map = {};
    
    for (let right = 0; right < s.length; right++) {
        while (s[right] in map) {
            delete map[s[left]];
            left += 1;
        }
        map[s[right]] = 1;
        longest = Math.max(longest, right - left + 1);
    }
    
    return longest;
    
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> seen = new HashSet<>();
        int left = 0;
        int right = 0;

        if (s.length() == 0) {
            return 0;
        }

        int longest = 0;

        char[] charArr = s.toCharArray();

        while (right < s.length()) {
            char currChar = charArr[right];
            if (!seen.contains(currChar)) {
                longest = Math.max(longest, right - left + 1);
            } else {
                while (seen.contains(currChar)) {
                    seen.remove(charArr[left]);
                    left++;
                }
            }
            right++;
            seen.add(currChar);
        }

        return longest;
    }
}